### PR TITLE
feat(grid): row selection templating #4998

### PR DESCRIPTION
**Pull request by jackofdiamond5**
Closes  https://github.com/IgniteUI/igniteui-angular/issues/4998

### Additional information (check all that apply):
 - [ ] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code
 - [x] This PR includes API docs for newly added methods/properties
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/grid-common.module.ts [hunk 1]
```diff
@@ -193,7 +194,8 @@ import { IgxRowDragModule } from './row-drag.directive';
         IgxGridPipesModule,
         IgxGridExcelStyleFilteringModule,
         IgxRowDragModule,
-        IgxPaginatorModule
+        IgxPaginatorModule,
+        IgxSelectorsModule
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/grid-common.module.ts]: you should add `IgxSelectorsModule` to `Exports` to make its entries visible

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 2]
```diff
@@ -259,6 +260,8 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
     private _locale = null;
     public _destroyed = false;
     private overlayIDs = [];
+    public rowSelected = false;
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: It seems this is not needed

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 3]
```diff
@@ -1707,9 +1710,53 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+     */
+    public get rowSelectorTemplate(): TemplateRef<IgxRowSelectorDirective> {
+        if (this.rowSelectorsTemplate && this.rowSelectorsTemplate.first) {
+            return this.rowSelectorsTemplate.first.templateRef;
+        }
+
+        return null;
+    }
+
+    @ContentChildren(IgxRowSelectorDirective, { read: IgxRowSelectorDirective, descendants: false })
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Add @hidden, @internal

File: projects/igniteui-angular/src/lib/grids/igx-selection.module.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+})
+export class IgxHeadSelectorDirective {
+    constructor(public templateRef: TemplateRef<any>) { }
+}
+
+@NgModule({
+    declarations: [IgxRowSelectorDirective, IgxHeadSelectorDirective],
+    exports: [IgxRowSelectorDirective, IgxHeadSelectorDirective]
+})
+export class IgxSelectorsModule {
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/grids/igx-selection.module.ts]: `IgxSelectorsModule` is too general as name, can it be `igxRowSelectorsModule`?

File: projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-grid-base.component.ts [hunk 5]
```diff
@@ -25,6 +26,8 @@ import { IgxHierarchicalGridNavigationService } from './hierarchical-grid-naviga
 import { IgxGridSummaryService } from '../summaries/grid-summary.service';
 import { IgxGridSelectionService, IgxGridCRUDService } from '../../core/grid-selection';
 import { IgxChildGridRowComponent } from './child-grid-row.component';
+import { IgxRowSelectorDirective, IgxHeadSelectorDirective } from '../igx-selection.module';
```
Comment (Lipata) [inline projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-grid-base.component.ts]: not used anymore

File: CHANGELOG.md [hunk 6]
```diff
@@ -32,6 +32,7 @@ For more information about the theming please read our [documentation](https://w
     - `uniqueColumnValuesStrategy` input is added. This property provides a callback for loading unique column values on demand. If this property is provided, the unique values it generates will be used by the Excel Style Filtering (instead of using the unique values from the data that is bound to the grid).
     - `igxExcelStyleLoading` directive is added, which can be used to provide a custom loading template for the Excel Style Filtering. If this property is not provided, a default loading template will be used instead.
     - introduced new properties `cellSelection` and `rowSelection` which accept GridSelection mode enumeration. Grid selection mode could be none, single or multiple. Also `hideRowSelectors` property is added, which allows you to show and hide row selectors when row selection is enabled.
+    - introduced functionality for templating row and header selectors
```
Comment (Lipata) [inline CHANGELOG.md]: Here you can extend with at least saying which directives can be used for templating. @damyanpetev I think we can also add code snippet showing how to do this?

Comment (damyanpetev) [reply CHANGELOG.md]: A short-ish snippet should be good. Also a link to readme or docs would be nice.

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 7]
```diff
@@ -1707,9 +1709,51 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+
+        return null;
+    }
+
+    /**
+     * @hidden
+     * @internal
+     */
+    @ContentChildren(IgxHeadSelectorDirective, { read: IgxHeadSelectorDirective, descendants: false })
+    public headSelectorsTemplate: QueryList<IgxHeadSelectorDirective>;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: this should be `headSelectorsTemplates`

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 8]
```diff
@@ -1707,9 +1709,51 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
...
+
+        return null;
+    }
+
+    /**
+     * @hidden
+     * @internal
+     */
+    @ContentChildren(IgxRowSelectorDirective, { read: IgxRowSelectorDirective, descendants: false })
+    public rowSelectorsTemplate: QueryList<IgxRowSelectorDirective>;
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: this should be `rowSelectorsTemplates`

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.html [hunk 9]
```diff
@@ -237,5 +242,17 @@
     <igx-icon fontSet="material">drag_indicator</igx-icon>
 </ng-template>
 
+<ng-template #headSelectorBaseTemplate>
+    <igx-checkbox 
+        [checked]="selectionService.areAllRowSelected()"
+        disableRipple="true"
+        [ngStyle]="{'visibility': isMultiRowSelectionEnabled? 'visible' : 'hidden' }"
+        (click)="$event.preventDefault()"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.html]: use read-only instead of this. IgxCheckBox read-only is already in master :)

File: projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-grid.component.html [hunk 10]
```diff
@@ -203,5 +210,17 @@
     <igx-icon fontSet="material">drag_indicator</igx-icon>
 </ng-template>
 
+<ng-template #headSelectorBaseTemplate>
+    <igx-checkbox
+        [checked]="selectionService.areAllRowSelected()"
+        disableRipple="true"
+        [ngStyle]="{'visibility': isMultiRowSelectionEnabled? 'visible' : 'hidden' }"
+        (click)="$event.preventDefault()"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-grid.component.html]: use read-only

File: projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-row.component.html [hunk 11]
```diff
@@ -56,3 +64,13 @@
     </igx-hierarchical-grid-cell>
 </ng-template>
 
+<ng-template #rowSelectorBaseTemplate>
+    <igx-checkbox
+        (click)="$event.preventDefault()"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-row.component.html]: use read-only for the checkbox

File: projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-row.component.html [hunk 12]
```diff
@@ -106,3 +109,13 @@
     </ng-template>
     <ng-container *ngTemplateOutlet="col.visibleIndex === 0 ? treeCellTemplate : cellTemplate"></ng-container>
 </ng-template>
+<ng-template #rowSelectorBaseTemplate>
+    <igx-checkbox
+        (click)="$event.preventDefault()"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-row.component.html]: use read-only

File: projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid.component.html [hunk 13]
```diff
@@ -187,5 +191,17 @@
     <igx-icon fontSet="material">drag_indicator</igx-icon>
 </ng-template>
 
+<ng-template #headSelectorBaseTemplate>
+    <igx-checkbox
+        [checked]="selectionService.areAllRowSelected()"
+        [ngStyle]="{'visibility': isMultiRowSelectionEnabled? 'visible' : 'hidden' }"
+        (click)="$event.preventDefault()"
```
Comment (wnvko) [inline projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid.component.html]: use read-only

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.html [hunk 14]
```diff
@@ -117,7 +122,7 @@
             [igxForItemSize]="hasColumnLayouts ? rowHeight * multiRowLayoutRowSize + 1 : renderedRowHeight"
             #verticalScrollContainer (onChunkPreload)="dataLoading($event)">
             <ng-template #record_template>
-                <igx-grid-row [gridID]="id" [index]="rowIndex" [rowData]="rowData" #row>
+                <igx-grid-row [gridID]="id" [index]="getGridRowIndex(rowData, rowIndex)" [rowData]="rowData" #row>
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.html]: Oh, this can have way more impact than you'd imagine (see tests).
Keep the different index inside the row template outlets context only!

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.html [hunk 15]
```diff
@@ -237,5 +242,17 @@
     <igx-icon fontSet="material">drag_indicator</igx-icon>
 </ng-template>
 
+<ng-template #headSelectorBaseTemplate igxHeadSelector let-context>
+    <igx-checkbox 
+        [readonly]="true"
+        [checked]="context.totalCount === context.selectedCount"
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.html]: This may need a additional `context.selectedCount &&` check in case there's no data? I assume those would be 0-s then?

File: projects/igniteui-angular/src/lib/grids/grid-base.component.ts [hunk 16]
```diff
@@ -1761,8 +1805,8 @@ export abstract class IgxGridBaseComponent extends DisplayDensityBase implements
     /**
      * @hidden
      */
-    @ViewChild('headerCheckbox', { read: IgxCheckboxComponent, static: false })
-    public headerCheckbox: IgxCheckboxComponent;
+    @ViewChild(IgxCheckboxComponent, { read: IgxCheckboxComponent, static: false })
```
[the remark this answers is not in the document]
Comment (damyanpetev) [reply projects/igniteui-angular/src/lib/grids/grid-base.component.ts]: Erm this is still here

File: projects/igniteui-angular/src/lib/grids/row.component.ts [hunk 17]
```diff
@@ -303,7 +302,7 @@ export class IgxRowComponent<T extends IgxGridBaseComponent & IGridDataBindable>
      * This method emits `onEditDone` event.
      *
      * ```typescript
-     * // update the second selected row's value
+    *  // update the second selected row's value
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/row.component.ts]: That didn't quite work.
```suggestion
     * // update the second selected row's value
```

File: projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-row.component.html [hunk 18]
```diff
@@ -106,3 +109,13 @@
...
+<ng-template #rowSelectorBaseTemplate>
+    <igx-checkbox
+        [readonly]="true"
+        [checked]="selected"
+        [disabled]="deleted"
+        disableRipple="true"
+        [disableTransitions]="grid.disableTransitions"
+        [aria-label]="rowCheckboxAriaLabel">
+    </igx-checkbox>
+</ng-template>
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/tree-grid/tree-grid-row.component.html]: New line

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.html (hunk 15 above)
Comment (damyanpetev) [reply projects/igniteui-angular/src/lib/grids/grid/grid.component.html]: image

File: projects/igniteui-angular/src/lib/grids/grid/grid.component.html [hunk 19]
```diff
@@ -58,17 +58,22 @@
...
                     'igx-grid__cbx-selection--push': filteringService.isFilterRowVisible
                 }">
-                    <igx-checkbox [checked]="selectionService.areAllRowSelected()" disableRipple="true" [ngStyle]="{'visibility': isMultiRowSelectionEnabled? 'visible' : 'hidden' }"
-                    (click)="$event.preventDefault()" [indeterminate]="selectionService.hasSomeRowSelected()" [aria-label]="headerCheckboxAriaLabel"  #headerCheckbox></igx-checkbox>
+                    <ng-template #headSelector
+                        *ngTemplateOutlet="
+                        this.headSelectorTemplate ? this.headSelectorTemplate : headSelectorBaseTemplate;
+                        context: { $implicit: {
+                                        selectedCount: this.selectionService.filteredAndSelectedData.length,
+                                        totalCount: this.filteredData ? this.filteredData.length : this.data.length }}">
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/grid/grid.component.html]: Bonus: Try with `[data]="null"` or don't set it at all. Filtered data won't be defined but the same is true for data, so either `.length` access will throw.

File: projects/igniteui-angular/src/lib/grids/grid/grid-row.component.html [hunk 20]
```diff
@@ -6,9 +6,15 @@
...
     </div>
 </ng-container>
-<ng-container *ngIf="showRowCheckboxes" >
+<ng-container *ngIf="this.showRowSelectors">
     <div class="igx-grid__cbx-selection" (click)="onRowSelectorClick($event)">
-        <igx-checkbox [checked]="selected" (click)="$event.preventDefault()" [disabled]="deleted"  disableRipple="true" [disableTransitions]="grid.disableTransitions" [aria-label]="rowCheckboxAriaLabel"></igx-checkbox>
+         <ng-template *ngTemplateOutlet="
+            this.grid.rowSelectorTemplate ? this.grid.rowSelectorTemplate : rowSelectorBaseTemplate;
+            context: { $implicit: {
+                            index: (this.grid.groupingExpressions ? this.grid.filteredSortedData.indexOf(this.rowData) : this.index) +
```
Comment (damyanpetev) [inline projects/igniteui-angular/src/lib/grids/grid/grid-row.component.html]: Alright, this can be moved to a getter in a separate PR.
